# Problem 1 - Extending the Lady Tasting Tea Experiment

The aim of this notebook, is to extend the classical *Lady Tasting Tea* experiment, as demonstrated in the module, to use a larger number of cups.

To achieve this, I will:

- Set up the original 8-cup experiment
- Set up extended 12-cup experiment,
- Use Numpy, to simulate both the the original 8 cup and extended 12-cup experiment
- Compare the simulated probabilities
- Interpret the results

# Importing Required Libraries

In [128]:
# Mathematical functions from the standard library.
# https://docs.python.org/3/library/math.html
import math

# Permutations and combinations.
# https://docs.python.org/3/library/itertools.html
import itertools

# Random selections.
# https://docs.python.org/3/library/random.html
import random

# Numerical structures and operations.
# https://numpy.org/doc/stable/reference/index.html#reference
import numpy as np

# Plotting.
# https://matplotlib.org/stable/contents.html
import matplotlib.pyplot as plt

import pandas as pd

from math import comb

# Experiment Set Up - Original and Extended

## Setup for Both Experiments

The designs of both the Experiments are detailed below:

1. **Original experiment as performed by Fisher:**
   - 8 cups in total
   - 4 cups with milk poured first
   - 4 cups with tea poured first

2. **Extended experiment  as carried out in this Notebook:**
   - 12 cups in total
   - 4 cups with milk poured first
   - 8 cups with tea poured first

For both experiments, the particiant is told there are exactly 4 milk-first cups and, in trn,
must try identify which 4 cups these are.

Under the null hypothesis (they possess no special ability to distuinguish between Milk First or Tea First), the participant is in essence
choosing 4 random cups at random from the total. The number of possible
guesses in each case is given by the binomial coefficient:

The expression $\binom{n}{k}$ represents the number of ways to choose $k$ cups from $n$ cups.


In the next two sections, I will set up both experiments and compute these potential combinations and probabilities of successfully identifying the four Milk First cups at random.


## Original Experiment - Cups of Tea

In [129]:
# --- Original 8-cup setup ---

original_num_cups = 8 # Setting the number of cups in the original experiment
original_num_milk_first = 4 # Setting the number of milk-first cups in the original experiment

original_num_comb = math.comb(original_num_cups, original_num_milk_first) # Calculating the number of combinations from the original experiment of choosing 4 cups from 8

# Exact probability of getting all 4 correct by chance.
original_prob_all_correct = 1 / original_num_comb # Calculating the exact probability of getting all 4 correct by chance in the original experiment

original_num_comb, original_prob_all_correct # Displaying the number of combinations and the exact probability of getting all 4 correct by chance in the original experiment

(70, 0.014285714285714285)

## Updated Experiment - Cups of Tea

In [130]:
# --- Extended 12-cup setup ---

ext_num_cups = 12 # Setting the number of cups in the extended experiment
ext_num_milk_first = 4 # Setting the number of milk-first cups in the extended experiment

ext_num_comb = math.comb(ext_num_cups, ext_num_milk_first) # Calculating the number of combinations from the extended setup of choosing 4 cups from 12

extended_prob_all_correct = 1 / ext_num_comb # Calculating the exact probability of getting all 4 correct by chance in the extended experiment

ext_num_comb, extended_prob_all_correct # Displaying the number of combinations and the exact probability of getting all 4 correct by chance in the extended experiment


(495, 0.00202020202020202)

# Running the Extended Experiment

## Identifying the number of Simulations to run using Monte Carlo Method

In [ ]:
num_of_iters = [1_000, 5_000, 10_000, 50_000, 100_000, 1_000_000, 2_000_000] # Defining different numbers of simulations to run for estimating the probability

results = [] # Creating an empty list to store the results

for n in num_of_iters: # Stating to run the below steps for each number of simulations in num_of_iters, where n is the number of simulations being run
    # Generate a number of random values, where the value is greater then/ equal to 0 and less than 1, from np.random.rand.
    # Using n samples allows us to run n independent Monte Carlo trials in a single, efficient step.
    # Each comparison (np.random.rand < extended_prob_all_correct) represents an independent Bernoulli trial with success probability equal to the true exact probability.
    # A “success” occurs when a value is less than the true exact probability (extended_prob_all_correct).
    # This comparison produces a Boolean array where True indicates a simulated correct full guess by chance. 
    # Taking the mean of this Boolean array gives the estimated probability based on these n simulated trials.
    estimate = (np.random.rand(n) < extended_prob_all_correct).mean()
    #Estimate the probability of getting all 4 correct by chance using n simulations.
    diff = estimate - extended_prob_all_correct # Calculating the difference between the estimate and the true probability
    results.append([n, estimate, diff]) # Appending the number of simulations, estimate, and difference to the results list

df_results = pd.DataFrame(results, columns=["Simulations", "Estimate", "Difference from Exact Probability"]) # Creating a DataFrame from the results list with specified column names

df_results # Displaying the results DataFrame

Simulations  Estimate  Difference from Exact Probability
0         1000  0.003000                           0.000980
1         5000  0.001400                          -0.000620
2        10000  0.002000                          -0.000020
3        50000  0.002560                           0.000540
4       100000  0.002000                          -0.000020
5      1000000  0.002029                           0.000009
6      2000000  0.002003                          -0.000017

## Using NumPy to run the Extended Experiment with 12 Cups (4 Milk First)

Again, under the null hypothesis, the participant is identifying the 4 Milk First cups at random from an extended number of 12 cups total.  

In order to estimate the probability of all 4 milk-first cups being correctly identified in the extended 12-cup experiment, we will run the experiment 100,000 times using NumPy. 

The simulation follows the Monte Carlo method described above, where each simulated run corresponds to an independent Bernoulli trial. 
The choice of 100,000 simulations is based on practical Monte Carlo considerations: using a large number of repetitions reduces sampling variability and ensures our estimate is close to the exact probability.

The exact probability is \(1/495\).  


### Running Extended Experiment

In [132]:
ext_milk_first_labels = set(range(ext_num_milk_first)) # Defining the Milk-First cup labels in the extended experiment

num_sims = 100000 # Setting the number of simulations to run based on Monte Carlo Simulation.

rng = np.random.default_rng() # Using NumPy's default random number generator to allow for the simulation to be repeatable

ext_num_of_successes = 0 # Setting the initial number of successes to zero

for i in range(num_sims): # Stating to run the below steps for number of simulations in num_sims
    # Randomly guessing which cups are milk-first without replacement.
    # Assigning the guessed milk-first cup labels to guessed_labels.
    # Using rng.choice to randomly select ext_num_milk_first unique cup labels from the total ext_num_cups.
    # ext_num_cups: Total number of cups (12 in this case).
    # size=ext_num_milk_first: Number of milk-first cups to guess (4 in this case).
    # replace=False: Ensures that the same cup label is not chosen more than once (no replacement).
    ext_guessed_labels = rng.choice(ext_num_cups, 
                                size=ext_num_milk_first,
                                replace=False)
    
    if set(ext_guessed_labels) == ext_milk_first_labels: # Checking if the guessed labels match the actual milk-first labels
        ext_num_of_successes += 1 # Incrementing the number of successes if the guess is correct

estimated_prob_all_correct_extended = ext_num_of_successes / num_sims # Calculating the estimated probability of getting all 4 correct by chance in the extended experiment based on the simulation

ext_num_of_successes, num_sims, estimated_prob_all_correct_extended # Displaying the number of successes, number of simulations, and the estimated probability from the simulation


(198, 100000, 0.00198)

### Running Simulation of Original Experiment

In [133]:
orig_milk_first_labels = set(range(original_num_milk_first)) # Defining the Milk-First cup labels in the original experiment


orig_num_of_successes = 0 # Setting the initial number of successes to zero

for i in range(num_sims): # Stating to run the below steps for number of simulations in num_sims
    # Randomly guessing which cups are milk-first without replacement.
    # Assigning the guessed milk-first cup labels to guessed_labels.
    # Using rng.choice to randomly select original_num_milk_first unique cup labels from the total original_num_cups.
    # original_num_cups: Total number of cups (8 in this case).
    # size=original_num_milk_first: Number of milk-first cups to guess (4 in this case).
    # replace=False: Ensures that the same cup label is not chosen more than once (no replacement).
    orig_guessed_labels = rng.choice(original_num_cups, 
                                size=original_num_milk_first,
                                replace=False)
    
    if set(orig_guessed_labels) == orig_milk_first_labels: # Checking if the guessed labels match the actual milk-first labels
        orig_num_of_successes += 1 # Incrementing the number of successes if the guess is correct

estimated_prob_all_correct_orig = orig_num_of_successes / num_sims # Calculating the estimated probability of getting all 4 correct by chance in the extended experiment based on the simulation

orig_num_of_successes, num_sims, estimated_prob_all_correct_orig # Displaying the number of successes, number of simulations, and the estimated probability from the simulation


(1435, 100000, 0.01435)

# Interpretation of Results

In [138]:


df_prob = pd.DataFrame({
    "Experiment": ["Original (8 cups)", "Extended (12 cups)"],
    "Total Cups": [8, 12],
    "Milk-First Cups": [4, 4],
    "Exact Probability": [original_prob_all_correct, extended_prob_all_correct],
    "Simulated Probability": [estimated_prob_all_correct_orig, estimated_prob_all_correct_extended],
})

df_prob


Experiment  Total Cups  Milk-First Cups  Exact Probability  \
0   Original (8 cups)           8                4           0.014286   
1  Extended (12 cups)          12                4           0.002020   

   Simulated Probability  
0                0.01435  
1                0.00198

## Interpreting the Estimated Probabilities

The table above shows both the exact and simulated probabilities of correctly identifying all milk-first cups purely by random guessing,that is, achieving a perfect score in each experiment.

### 2. Interpretation of the Results

- **Original 8-cup experiment:**  
  The exact probability of a perfect score is roughly 1/70 or 0.0143 for the Original 8 Cup Experiment.  
  This means that even with no genuine ability, a participant would, on average, guess all 4 milk-first cups correctly about 1.4% of the time.

- **Extended 12-cup experiment:**  
  The exact probability decreases to 1/495 or 0.0020 for the Extended 12 Cup Experiment 
  This means that even with no genuine ability, a participant would, on average, guess all 4 milk-first cups correctly about 0.2% of the time when the number of cups is increased to 12.

By increasing the total number of cups from 8 to 12, while keeping the number of milk-first cups fixed at 4, it makes it much harder for a random guesser to pick the correct combination. As a result, the probability of achieving a perfect score by chance becomes much smaller.

In fact, the liklihood of achieving a perfect score in the 12-cup version is over 7 times **less likely** than in the original experiment, under the assumption of Random Guessing.

---

### 3. Interpreting the p-values

Under the null hypothesis (they possess no special ability to distuinguish between Milk First or Tea First), the participant is in essence
choosing 4 cups at random from the all cups. A perfect score in this situation is therefore extremely unlikely, and the p-value tells us *how unlikely* it is.

In the scenario that we use a perfect score, as the criteria to detect ability:

- **Original 8-cup experiment:**  
  - The chance of a perfect guess is about 0.014 (1.4%).  
  - This is below the 5%/ 0.05 threshold level low enough to denote it as statistically significant.
    - This is turn, may indicate an innate ability to actually identify the cups that were Milk First.  
  - However, if we were to use the stricter 1%/ 0.01 threshold level, this would not be deemed to be statistically significant.

- **Extended 12-cup experiment:**  
  - The chance of a perfect guess is about 0.002 (0.2%), when the experiment is extended to have 12 cups.   
  - This is below both the 5% and 1% levels, meaning a perfect score provides very strong reasoning to believe a participant has an ability to tell the difference.

Based on the above findings, we would reject the null hypothesis in both experiments, but the extended version provides much stronger evidence, as the original experiment only meets the 5% significance level, while the extended experiment is also significant at the stricter 1% level.

As the extended experiment results in an extremely small p-value (≈ 0.002), it would still be considered significant even under the more stringent thresholds proposed by Benjamin & Berger (2019), who argued for lowering the traditional 0.05 standard. In the case of a stricter threshold, the original experiment, by contrast, would not meet it.

However, in standard practice, researchers typically retain the conventional 5% or 1% significance levels, in which case both experiments allow rejection of the null hypothesis, with the extended design providing much stronger evidence.
---



# References
1. Number of Simulations - https://statisticsbyjim.com/probability/monte-carlo-simulation/
2. Number of Simulations - https://www.statology.org/how-to-perform-monte-carlo-simulations-in-python-with-example/
3. Number of Simulations - https://www.statology.org/the-concise-guide-to-monte-carlo-simulation/
4. Numpy.Random - https://realpython.com/numpy-random-number-generator/
5. Numpy.Random - https://www.geeksforgeeks.org/python/numpy-random-rand-python/
6. Numpy.Random - https://numpy.org/doc/stable/reference/random/generated/numpy.random.rand.html
7. Numpy.Random - https://sparkbyexamples.com/python/python-numpy-random-rand-function/
8. Numpy.Random - https://www.programiz.com/python-programming/numpy/random
9. For Loops - https://www.geeksforgeeks.org/python/loops-in-python/
10. Bernouli Trials - https://www.geeksforgeeks.org/maths/bernoulli-trials-binomial-distribution/
11. Bernouli Trials - https://github.com/uros-bojanic/bernoulli-trials
12. Monte Carlo Simulation and Bernouli - https://medium.com/@polanitzer/monte-carlo-simulation-and-bernoulli-distribution-in-python-predict-next-years-tax-exemption-ec6f3d8b6ba9
13. Random Number Generator - https://realpython.com/numpy-random-number-generator/
14. Random.Choice - https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.choice.html
15. Random.Choice - https://www.geeksforgeeks.org/python/random-choices-method-in-python/
16. Markdown - https://www.geeksforgeeks.org/machine-learning/markdown-cell-in-jupyter-notebook/
17. Markdown - https://stackoverflow.com/questions/36288670/how-to-programmatically-generate-markdown-output-in-jupyter-notebooks
18. Markdown - https://www.datacamp.com/tutorial/markdown-in-jupyter-notebook
19. Markdown - https://www.markdownguide.org/basic-syntax/
20. p-Value Interpretation - https://statisticsbyjim.com/hypothesis-testing/interpreting-p-values/
21. p-Value Interpretation - https://www.geeksforgeeks.org/machine-learning/p-value/
22. Benjamin, D. J., & Berger, J. O. (2019). Three recommendations for improving the use of p-values. The American Statistician, 73(sup1), 186–191. https://doi.org/10.1080/00031305.2018.1543135